flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamTaskStateInitializerImpl.java

```diff
@@ -176,15 +180,31 @@ public StreamOperatorStateContext streamOperatorStateContext(
         try {
 
             // -------------- Keyed State Backend --------------
-            keyedStatedBackend =
-                    keyedStatedBackend(
-                            keySerializer,
-                            operatorIdentifierText,
-                            prioritizedOperatorSubtaskStates,
-                            streamTaskCloseableRegistry,
-                            metricGroup,
-                            managedMemoryFraction,
-                            statsCollector);
+            // TODO: Only init async keyed state backend if supported, consider to init different
+            // state backend by the operator type.
+            if (stateBackend.supportsAsyncKeyedStateBackend()) {
+                asyncKeyedStateBackend =
+                        keyedStatedBackend(
+                                keySerializer,
+                                operatorIdentifierText,
+                                prioritizedOperatorSubtaskStates,
+                                streamTaskCloseableRegistry,
+                                metricGroup,
+                                managedMemoryFraction,
+                                statsCollector,
+                                StateBackend::createAsyncKeyedStateBackend);
+            } else {
```

Comment:
Do you mean only create one type of state backend (sync/async). I'd suggest splitting this up, like https://github.com/apache/flink/pull/24697/files#diff-60f97b32726f78481f7852ea82daa2f3c9919933c93c234a049442195ccf0e2b does.

Reply:
As discussed offline, Since current creation of `KeyedStateBackend` contains some allocation of resources, e.g. managed memory, we remain creating one type here. 
Also add a TODO to support `KeyedStateBackend` for `AsyncKeyedStateBackend` in the future.